Data/Aeson/Types/ToJSON.hs

```diff
@@ -318,17 +318,102 @@ instance KeyValue Pair where
+--   {"5":"a"}
+--
+--   JSON documents only accept strings as object keys. For any type
+--   from @base@ that has a natural textual representation, it can be
+--   expected that its 'ToJSONKey' instance will choose that representation.
+--
+--   For data types that lack a natural textual representation, an alternative
+--   is provided. The map-like container is represented as a JSON array
+--   instead of a JSON object. Each value in the array is an array with
+--   exactly two values. The first is the key and the second is the value.
+--
+--   For example, values of type '[Text]' cannot be encoded to a
+--   string, so a 'Map' with keys of type '[Text]' is encoded as follows:
+--
+--   >>> LBC8.putStrLn $ encode $ Map.fromList [(["foo","bar","baz" :: Text], 'a')]
+--   [[["foo","bar","baz"],"a"]]
+--
+--   The default implementation of 'ToJSONKey' chooses this method of
+--   encoding a key, using the 'ToJSON' instance of the type.
+--
+--   To use your own data type as the key in a map, all that is needed
+--   is to write a 'ToJSONKey' (and possibly a 'FromJSONKey') instance
+--   for it. If the type cannot be trivially converted to and from 'Text',
+--   it is recommended that 'ToJSONKeyValue' is used. Since the default
+--   implementations of the typeclass methods can build this from a
+--   'ToJSON' instance, there is nothing that needs to be written:
+--
+--   > data Foo = Foo { fooAge :: Int, fooName :: Text }
+--   >   deriving (Eq,Ord,Generic)
+--   > instance ToJSON Foo
+--   > instance ToJSONKey Foo
+--
+--   That's it. We can now write:
+--
+--   >>> let m = Map.fromList [(Foo 4 "bar",'a'),(Foo 6 "arg",'b')]
+--   >>> LBC8.putStrLn $ encode m
+--   [[{"fooName":"bar","fooAge":4},"a"],[{"fooName":"arg","fooAge":6},"b"]]
+--   
+--   The next case to consider is if we have a type that is a
+--   newtype wrapper around 'Text'. The recommended approach is to use
+--   generalized newtype deriving:
+--
+--   > newtype RecordId = RecordId { getRecordId :: Text}
+--   >   deriving (Eq,Ord,ToJSONKey)
+--   
+--   Then we may write:
+--   
+--   >>> LBC8.putStrLn $ encode $ Map.fromList [(RecordId "abc",'a')]
+--   {"abc":"a"}
+--
+--   Simple sum types are a final case worth considering. Suppose we have:
+--
+--   > data Color = Red | Green | Blue | Black
+--   >   deriving (Show,Read,Eq,Ord)
+--
+--   It is possible to get the 'ToJSONKey' instance for free as we did
+--   with 'Foo'. However, in this case, we have a natural way to go to
+--   and from 'Text' that does not require any escape sequences. So, in
+--   this example, 'ToJSONKeyText' will be used instead of 'ToJSONKeyValue'.
+--   At the moment, there are no conveniences for this situation. The
+--   'Show' instance can be used to help write 'ToJSONKey', but this will 
+--   be inefficient since it uses 'String' as an intermediate step:
+--
+--   > instance ToJSONKey Color where
+--   >   toJSONKey = ToJSONKeyText f g
```

Comment:
I do have a [small benchmark](https://github.com/RyanGlScott/text-show/blob/b49cd9444bf8c057517f8ea7057bbc5bd6954615/benchmarks/Bench.hs) in `text-show` which compares the performance of calling `Text.pack . show` versus using `text-show`'s functionality. On a [recent benchmark run](https://travis-ci.org/RyanGlScott/text-show/jobs/136144077#L1537), I obtained the following results:

```
benchmarking String Show, then Text.pack/Small sample
time                 4.235 μs   (4.215 μs .. 4.257 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 4.245 μs   (4.221 μs .. 4.284 μs)
std dev              102.3 ns   (70.67 ns .. 135.3 ns)
variance introduced by outliers: 28% (moderately inflated)

benchmarking String Show, then Text.pack/Medium sample
time                 8.068 ms   (8.028 ms .. 8.130 ms)
                     0.999 R²   (0.996 R² .. 1.000 R²)
mean                 8.107 ms   (8.053 ms .. 8.187 ms)
std dev              185.4 μs   (96.48 μs .. 288.5 μs)

benchmarking String Show, then Text.pack/Large sample
time                 1.003 s    (736.8 ms .. 1.291 s)
                     0.990 R²   (0.964 R² .. 1.000 R²)
mean                 1.041 s    (1.009 s .. 1.072 s)
std dev              53.56 ms   (0.0 s .. 54.26 ms)
variance introduced by outliers: 19% (moderately inflated)

benchmarking Text Show (TH)/Small sample
time                 1.775 μs   (1.768 μs .. 1.785 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 1.827 μs   (1.801 μs .. 1.876 μs)
std dev              119.4 ns   (64.69 ns .. 194.0 ns)
variance introduced by outliers: 76% (severely inflated)

benchmarking Text Show (TH)/Medium sample
time                 3.094 ms   (3.037 ms .. 3.172 ms)
                     0.997 R²   (0.996 R² .. 0.999 R²)
mean                 3.102 ms   (3.071 ms .. 3.136 ms)
std dev              107.5 μs   (87.66 μs .. 129.3 μs)
variance introduced by outliers: 17% (moderately inflated)

benchmarking Text Show (TH)/Large sample
time                 360.0 ms   (346.2 ms .. 376.6 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 373.3 ms   (368.0 ms .. 378.0 ms)
std dev              7.535 ms   (67.99 as .. 8.053 ms)
variance introduced by outliers: 19% (moderately inflated)
```

So your hunch was correct: `text-show` is indeed more performant (at least in this scenario in the benchmark).